### PR TITLE
Add HTTP/2 Support

### PR DESCRIPTION
**Pull request by Varuni-Punchihewa**
## Purpose
The WSO2 API Microgateway currently supports only HTTP/1.1 version via HTTP and HTTPS as the incoming and outgoing transport protocols. This feature enables the HTTP/2 support for both incoming and outgoing message flows.

## Goals
- If the client enables the HTTP/2 feature, the microgateway sets its maximum httpVersion to 2.0
- When client sends an HTTP/2 requests the microgateway upgrades the connection between the client and the microgateway to h2, and if the BE supports HTTP/2 the microgateway upgrades the connection between the microgateway and the BE to h2.

## Approach
The gateway server is written in Ballerina (version 0.981.1) and in order to give the HTTP/2 support, we have to change the HTTP version supported by the client and listener endpoints into 2.0 as stated below. 

```ballerina
endpoint http:Listener testEP {
port: 9095,
httpVersion: "2.0"
};
```

When the user Setup the microgateway, it will generate the source files, “listeners.bal and “service.bal” where the gateway would read from when it runs. 

For the gateway to identify whether the user wants to enable the HTTP/2 feature, the user is required to change the http2Enable into true which is found in the micro-gw.conf file generated after building the microgateway.

When we run the gateway, it needs to set the http version into 2.0 according to what is provided in the micro-gw.conf. This is achieved by introducing a new http2.mustache file and including the following line of code.

```ballerina
,
    httpVersion: gateway:getHttpVersion()
```

getHttpVersion() method is implemented in ballerina/gateway/http2.bal file as below

```ballerina
import ballerina/io;
import ballerina/config;

public function getHttpVersion() returns string {
    boolean val = config:getAsBoolean("http2.http2Enable");
    if(val) {
        return "2.0";
    } else {
        return "1.1";
    }
}
```

The http2.mustache is then included in httpEndpoint.mustache inside the Client endpoint and in listeners.mustache inside the APIGatewaySecureListener and APIGatewayListener endpoints.

In the api_gateway_listener.bal also the httpVersion is set to 2.0 as below if the user has changed the micro-gw.conf to enable the HTTP/2 support.

```ballerina
if (getConfigBooleanValue(HTTP2_INSTANCE_ID, HTTP2_PROPERTY, false)) {
        config.httpVersion = "2.0";
    }
 ```
, where HTTP2_INSTANCE_ID = http2, HTTP2_PROPERTY = http2Enable.

## User stories

scene1-1

| User Story  | Both the client and the backend support HTTP/2 |
| ------------- | ------------- |
| Overview  | The client sends an HTTP/2 request to the backend via Microgateway. The HTTP/2 request is first received by the Microgateway, and then it forwards it to the HTTP/2 supported backend |
| Prerequisites  | <ul><li>User setup a Microgateway project </li> <li>The user builds the Microgateway project </li><li>User goes into the built project, change the http2Enable into true (http2Enable=true) in the micro-gw.conf file</li> <li>Run the Gateway </li></ul> |
| Narrative | <ul><li>The user sends an HTTP/1.1 request with the HTTP/2 headers to the Microgateway</li><li>The microgateway accepts the request with the headers and sends a response back to the client saying the connection is upgraded to h2</li><li>The microgateway forwards the received HTTP/1.1 request with the h2 headers to the backend</li><li>The backend first sends a response back to the microgateway saying the connection is upgraded to h2 and after that it sends an HTTP/2 response with the relevant data back to the Microgateway</li><li>The Microgateway forwards that HTTP/2 response back to the client</li></ul>
| Acceptance criteria | For the HTTP/2 request that the client has sent, he will receive an HTTP/2 response |

scene2-1

| User Story  | The client supports HTTP/2 but the backend does not support HTTP/2 |
| ------------- | ------------- |
| Overview  | The client sends an HTTP/2 request to the Backend via Microgateway. The HTTP/2 request is first received by the Microgateway, and then it forwards it to the Backend |
| Prerequisites  | <ul><li>User setup a Microgateway project</li><li>The user builds the Microgateway project</li><li>User goes into the built project, change the http2Enable into true (http2Enable=true) in the micro-gw.conf file</li><li>Run the Gateway</li></ul> |
| Narrative | <ul><li>The user sends an HTTP/1.1 request with the HTTP/2 headers to the Microgateway</li><li>The Microgateway accepts the request with the headers and sends a response back to the client saying the connection is upgraded to h2</li><li>The microgateway then forwards the received HTTP/1.1 request  with the h2 headers to the backend</li><li>Since the backend does not support the h2, it does not upgrade the connection into h2 and it sends an HTTP/1.1 response back to the Microgateway</li><li>Microgateway then upgrades it to an HTTP/2 response by adding h2 headers and sends it back to the client</li></ul> |
| Acceptance criteria | For the HTTP/2 request that the client has sent, he will receive an HTTP/2 response |

scene3-1

| User Story  | The client does not support HTTP/2 but the backend supports HTTP/2 |
| ------------- | ------------- |
| Overview  |  The client sends an HTTP/1.1 request to the backend via Microgateway. The HTTP/1.1 request is first received by the Microgateway, and forwards it to the HTTP/2 supported backend  |
|  Prerequisites  |  <ul><li>User setup a Microgateway projectThe user builds the Microgateway project</li><li>User goes into the built project, change the http2Enable into true (http2Enable=true) in the micro-gw.conf file Run the Gateway</li><li>The user sends an HTTP/1.1 request to the Microgateway</li><li>The Microgateway upgrades it to an HTTP/2 request by adding h2 headers and sends it to the backend</li><li>The backend first sends a response back to the microgateway saying the connection is upgraded to h2 and after that it sends an HTTP/2 response with the relevant data back to the Microgateway</li><li>The Microgateway then downgrades it to an HTTP/1.1 response removing the h2 headers and sends it back to the client</li></ul> |
| Narrative | <ul><li>The user sends an HTTP/1.1 request to the Microgateway</li><li>The Microgateway upgrades it to an HTTP/2 request by adding h2 headers and sends it to the backend</li><li>The backend first sends a response back to the microgateway saying the connection is upgraded to h2 and after that it sends an HTTP/2 response with the relevant data back to the Microgateway</li><li>The Microgateway then downgrades it to an HTTP/1.1 response removing the h2 headers and sends it back to the client</li><ul> |
| Acceptance criteria | For the HTTP/1.1 request that the client has sent, he will receive an HTTP/1.1 response |

## Documentation
[https://docs.google.com/document/d/1H4_-IfyPcC1UcIgWNwFnnYRuVFo-TKPB2lUx3we0t74/edit?usp=sharing](url)

## Automation tests
 - Unit tests 
1. testHTTP2RequestsViaInsecureConnectionWithHTTP1BE
2. testHTTP2RequestsViaSecureConnectionWithHTTP1BE
3. testHTTP2RequestsViaInsecureConnectionWithHTTP2BE
4. testHTTP2RequestsViaSecureConnectionWithHTTP2BE
5. testHTTP1RequestsViaInsecureConnectionWithHTTP2BE
6. testHTTP1RequestsViaSecureConnectionWithHTTP2BE

 - Integration tests

1. HTTP2RequestsWithHTTP1BackEndTestCase
2. HTTP2RequestsWithHTTP2BackEndTestCase

### CODE REVIEW

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java [hunk 1]
```diff
@@ -30,12 +30,7 @@
 import org.wso2.apimgt.gateway.cli.config.TOMLConfigParser;
 import org.wso2.apimgt.gateway.cli.constants.GatewayCliConstants;
 import org.wso2.apimgt.gateway.cli.constants.RESTServiceConstants;
-import org.wso2.apimgt.gateway.cli.exception.BallerinaServiceGenException;
-import org.wso2.apimgt.gateway.cli.exception.CLIInternalException;
-import org.wso2.apimgt.gateway.cli.exception.CLIRuntimeException;
-import org.wso2.apimgt.gateway.cli.exception.CliLauncherException;
-import org.wso2.apimgt.gateway.cli.exception.ConfigParserException;
-import org.wso2.apimgt.gateway.cli.exception.HashingException;
+import org.wso2.apimgt.gateway.cli.exception.*;
```
Comment (prasa7) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: remove the start imports

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java [hunk 2]
```diff
@@ -22,11 +22,7 @@
 import org.wso2.apimgt.gateway.cli.exception.BallerinaServiceGenException;
 import org.wso2.carbon.apimgt.rest.api.publisher.dto.APIInfoDTO;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
+import java.util.*;
```
Comment (prasa7) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java]: Remove the star imports

File: components/micro-gateway-core/src/main/ballerina/gateway/listeners/api_gateway_listener.bal [hunk 3]
```diff
@@ -131,6 +131,12 @@ function initiateGatewayConfigurations(EndpointConfiguration config) {
     initGatewayCaches();
     printDebug(KEY_GW_LISTNER, "Initialized gateway caches");
     initializeAnalytics();
+
+    //Change the version of http2
+    if(getConfigBooleanValue(HTTP2_INSTANCE_ID, HTTP2_PROPERTY, false)) {
```
Comment (mushthaq33) [inline components/micro-gateway-core/src/main/ballerina/gateway/listeners/api_gateway_listener.bal]: Lets format this.

File: components/micro-gateway-core/src/main/ballerina/gateway/http2.bal [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import ballerina/io;
+import ballerina/config;
+
+public function getHttpVersion() returns string {
+    boolean val = config:getAsBoolean("http2.http2Enable");
+    if(val) {
```
Comment (mushthaq33) [inline components/micro-gateway-core/src/main/ballerina/gateway/http2.bal]: Lets format this.

**Pull request merged**